pkgs/development/python-modules/gradio/default.nix

```diff
@@ -0,0 +1,204 @@
+  src = fetchPypi {
+    inherit pname version;
+    hash = "sha256-oG97GwehyBWjWXzDqyfj+x2mAfM6OQhYKdA3j0Rv8Vs=";
+  };
+
+  postPatch = ''
+    substituteInPlace requirements.txt \
+      --replace \
+        "mdit-py-plugins<=0.3.3" \
+        "mdit-py-plugins"
+  '';
+
+  nativeBuildInputs = [
+    hatchling
+    hatch-requirements-txt
+    hatch-fancy-pypi-readme
+  ];
+
+  propagatedBuildInputs = [
+    aiofiles
+    aiohttp
+    altair
+    fastapi
+    ffmpy
+    markdown-it-py
+    mdit-py-plugins
+    markupsafe
+    matplotlib
+    numpy
+    orjson
+    pandas
+    pillow
+    pycryptodome
+    python-multipart
+    pydub
+    pyyaml
+    requests
+    uvicorn
+    jinja2
+    fsspec
+    httpx
+    pydantic
+    websockets
+    typing-extensions
+  ] ++ markdown-it-py.optional-dependencies.linkify;
+
+  nativeCheckInputs = [
+    pytestCheckHook
+    pytest-asyncio
+    mlflow
+    #cometml # FIXME: enable once comet-ml is packaged
+    huggingface-hub
+    transformers
+    wandb
+    respx
+    scikit-image
+    ipython
+    ffmpeg
+    vega_datasets
+    boto3
+   ] ++ lib.optional shapWorks shap;
+
+  # Add pytest hook skipping tests that access network, marking them as "Expected fail" (xfail)
+  # We additionally xfail FileNotFoundError, as gradio often fail to upload test files to pypi.
+  preCheck = let
+    conftestSkipNetworkErrors = writeText "conftest-network-xfail.py" ''
+      from _pytest.runner import pytest_runtest_makereport as orig_pytest_runtest_makereport
+      import urllib, urllib3, httpx, requests, websockets
+
+      class NixNetworkAccessDeniedError(BaseException): pass
+      def deny_network_access(*a, **kw):
+        raise NixNetworkAccessDeniedError
+
+      def iterate_exc_chain(exc: Exception):
+        yield exc
+        if getattr(exc, "__context__", None) is not None:
+          yield from iterate_exc_chain(exc.__context__)
+        if getattr(exc, "__cause__", None) is not None:
+          yield from iterate_exc_chain(exc.__cause__)
+
+      httpx.Request = \
+      httpx.AsyncClient.get = \
+      httpx.AsyncClient.head = \
+      requests.head = \
+      requests.get = \
+      requests.post = \
+      urllib.request.urlopen = \
+      urllib.request.Request = \
+      urllib3.connection.HTTPSConnection._new_conn = \
+      websockets.connect = \
+        deny_network_access
+
+      def pytest_runtest_makereport(item, call):
+        tr = orig_pytest_runtest_makereport(item, call)
+        if call.excinfo is not None:
+          for exc in iterate_exc_chain(call.excinfo.value):
+            if isinstance(exc, NixNetworkAccessDeniedError):
+              tr.outcome, tr.wasxfail = 'skipped', "reason: Requires network access."
+            if isinstance(exc, FileNotFoundError):
+              tr.outcome, tr.wasxfail = 'skipped', "reason: Pypi dist bad."
+        return tr
+    '';
+  in ''
+    export HOME=$TMPDIR
+    cat ${conftestSkipNetworkErrors} >> test/conftest.py
+  '';
```

Comment:
Is some of that upstreamable? Like a `network` marker? As I mentioned, this looks annoying to maintain.

Reply:
IMO this is too related to nix to matter upstream.
How about we keep this as is, and if for some reason this doesn't work anymore, we just disable the tests until someone has time to look into this.
Since there are quite a lot of packages depending on gradio, would you rather 
- disable the tests until we confirm with upstream if they are interested in this or not.
- leave this as is for now and disable the tests if there is a failure related to this.

Reply:
> IMO this is too related to nix to matter upstream.

Having an extra marker to skip network tests is something not nix specific, many distros build their packages without networking. It can also be a great way to skip tests if no network is available or only very slow and flaky.

Reply:
In the prior pr I discussed generalizing this into a pytestCheckHook-esqe hook. The sentiment then was first pushing through this pr

Reply:
If the usecase comes up, we can always move the implementation into a hook but I am not fully convinced that this is necessary here.

What's the problem with upstreaming this and adding a marker to skip network tests? We just have a build environment without internet access but we would still like to run as many tests as possible.

Reply:
I don't trust the gradio devs, who tend to move fast and loose, to support this use case. My reason is that i did [this pr](https://github.com/gradio-app/gradio/pull/1706), to fix the missing test files on pypi. They they promptly broke it again without a few weeks.

I have bumped the version of the prior gradio pr about 3 times over the past year, and each time the list of tests to disable needed change. This proved quite time consuming  to maintain (hence that debug helper comment). Network and non-network tests are not grouped in any meaningful sense, so `disabledTestPaths` proved too aggressive while `disabledTests` proved too surgical and unmaintainable.

The conftest network xfail approach came upon me in a dream as an easily generalizable drop-in way to run as many tests as possible, without maintaining a `disabledTests` list, nor relying on upstream devs to support us poor distro maintainers. It is my brain child and I will defend it with passion :P.

To argue in favor of generalizing this xfail approach and a future check hook, here is a very coarse estimate of the number of packages which disable network tests, and likely a lot more, with `disableTestPaths`:

```shell
$ rg disabledTestPaths pkgs/development/python-modules/ -l | xargs rg network | wc -l
137
```

The conftest network xfail approach will for these likely improve testing coverage by a lot, and I believe such a hook will even remove a lot of `doChecks = false;` found in throughout our python package set.